include/ensmallen_bits/callbacks/early_stop_at_min_loss.hpp

```diff
@@ -28,11 +30,35 @@ class EarlyStopAtMinLoss
    * @param patienceIn The number of epochs to wait after the minimum loss has
    *    been reached or no improvement has been made (Default: 10).
    */
-  EarlyStopAtMinLoss(const size_t patienceIn = 10) :
-      patience(patienceIn),
-      bestObjective(std::numeric_limits<double>::max()),
-      steps(0)
-  { /* Nothing to do here */ }
+  EarlyStopAtMinLoss<AnnType>(const size_t patienceIn = 10,
+                              std::ostream& output = arma::get_cout_stream())
+    : patience(patienceIn), bestObjective(std::numeric_limits<double>::max()),
+      steps(0), output(output)
+  { /* Nothing to do here */
+  }
+
+  /**
+   * Set up the early stop at min class, which keeps track
+   * of the minimum loss and stops the optimization process if the loss stops
+   * decreasing on a specific validation set.
+   *
+   * @param predictors: data matrix used to predict the responses.
+   * @param responses: data matrix used to evaluate the predictions.
+   * @param patienceIn The number of epochs to wait after the minimum loss has
+   * been reached or no improvement has been made (Default: 10).
+   */
+  EarlyStopAtMinLoss<AnnType>(AnnType& network,
```

Comment:
I guess since this can be used with other methods as well, we should rename the parameter, what about `function`?

Reply:
Right, seems like maybe this would be a perfect use case for a `std::function`?  We could be really general about it.  The user can just write a lambda that, when called, evaluates the loss on the given parameters...

```
// assume we built this network somehow and we want to optimize it...
FFN<...> network;
// we have some test set and responses...
arma::mat testSet;
arma::mat testResponses;

EarlyStopAtMinLoss e(
    [network, testSet, testResponses](const arma::mat& /* parameters */) -> double
    {
      return network.Evaluate(testSet, testResponses);
    });

network.Train(trainSet, trainResponses, e, ...);
```

That's just a sketch, but that's sufficiently general that the user can do whatever they want with it, with the easiest example being using a validation/test set to compute a loss.  I guess, from my end, I want to make sure that the callback isn't specific to the machine learning setting (since ensmallen is more general, it's just generic optimization).

Reply:
I like the idea.

Reply:
@shrit what do you think, would it be better to take any `std::function` here instead of requiring something that has an `Evaluate()` method?

Reply:
@rcurtin Maybe I can add another constructor that takes an `std::function`, at the same time the user might do anything in `std::function` and not only `Evaluate`
I am not sure I can replace everything by `std::function`, since the `network.Evaluate` need to be called at the end of each epoch. right? I am not sure

Reply:
Yeah, adding another constructor could be a nice thing to do, so that the technique is more general.  I guess, my hope is not to link the mlpack and ensmallen codebases too closely here, so that people can use this callback in non-mlpack situations.  If the user passes a lambda too, that actually makes the implementation easier, since `EarlyStopAtMinLoss` doesn't have to then track the validation set and responses (it will be implicitly held in the given lambda/`std::function`).